### PR TITLE
Adding GyroDropout as an algorithm to Composer

### PR DESCRIPTION
**Pull request by jelite**
# What does this PR do?
This PR adds an algorithm gyro_dropout, including a simple example - gyro_dropout_example.py in composer/examples directory.

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/algorithms/gyro_dropout/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+### Implementation Details
+
+Gyro Dropout is implemented by performing model surgery, which looks for instances of `torch.nn.Dropout`. This should be applicable to any model that utilizes a `torch.nn.Dropout`.
+
+## Suggested Hyperparameters
+
+Gyro Dropout has two hyperparameters - sigma, tau. (iters_per_epoch and max_epoch is training-dependent)
+
+We recommend (256, 16) or (1024, 8) as hyperparameter set for (sigma, tau).
```
Comment (dskhudia) [inline composer/algorithms/gyro_dropout/README.md:72]: Which model(s) are these hyperparameters suitable for?

File: composer/algorithms/gyro_dropout/gyro_dropout.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    @staticmethod
+    def required_on_load() -> bool:
+        return True
+
+    def match(self, event: Event, state: State) -> bool: # XXX: not sure if this is correct.
+        del state  # unused
+        return event == Event.INIT
+
+    def apply(self, event: Event, state: State, logger: Logger) -> Optional[int]:
+        del event, logger  # unused   XXX: not sure if this is correct. Copied from FusedLayerNorm.
```
Comment (dskhudia) [inline composer/algorithms/gyro_dropout/gyro_dropout.py:137]: This is just needed to silence some lint warnings. You can remove the comment.

File: composer/algorithms/gyro_dropout/gyro_dropout.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def from_Dropout(p: float, sigma: int, tau: int, num_iterations: int, epoch:int, layer: torch.nn.Module, module_index: int):
+    """Defines a replacement policy from a `torch.nn.Dropout` to a 'GyroDropout`"""
+
+    return GyroDropoutLayer(p, sigma, tau, num_iterations, epoch)
+
+
+def apply_gyro_dropout(model: torch.nn.Module, 
+                       iters_per_epoch: int,
```
Comment (dskhudia) [inline composer/algorithms/gyro_dropout/gyro_dropout.py:66]: Algorithms being dependent on iters_per_epoch and max_epoch is not ideal as we will have to change these parameters for this algorithm manually whenever training duration changes. You can access training progress from state. cc: @mvpatel2000

File: composer/algorithms/gyro_dropout/gyro_dropout.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    def apply(self, event: Event, state: State, logger: Logger) -> Optional[int]:
+        del event, logger          
+        
+        apply_gyro_dropout(model=state.model,
+                           iters_per_epoch=int(state.dataloader_len),
+                           max_epoch=int(state.max_duration),
+                           p=self.p,
+                           sigma=self.sigma,
+                           tau=self.tau,
+        )
```
Comment (jelite) [inline composer/algorithms/gyro_dropout/gyro_dropout.py:143]: We deleted parameters -(iters_epoch, max_epoch) by using state. It works in composer's trainer, but user should write iters_per_epoch, max_epoch as parameters in functional api because functional api doesn't support state.

File: composer/algorithms/gyro_dropout/README.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+### Implementation Details
+
+Gyro Dropout is implemented by performing model surgery, which looks for instances of `torch.nn.Dropout`. This should be applicable to any model that utilizes a `torch.nn.Dropout`.
+
+## Suggested Hyperparameters
+
+Gyro Dropout has two hyperparameters - sigma, tau. (iters_per_epoch and max_epoch is training-dependent)
+
+We recommend (256, 16) in AlexNet, LeNet or (1024, 8) in ResNet-18, BERT as hyperparameter set for (sigma, tau).
```
Comment (jelite) [inline composer/algorithms/gyro_dropout/README.md:70]: Added model names that matches given parameters.

File: composer/algorithms/gyro_dropout/README.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+
+[\[How to Use\]](#how-to-use) - [\[Suggested Hyperparameters\]](#suggested-hyperparameters) - [\[Technical Details\]](#technical-details) - [\[Attribution\]](#attribution)
+
+ `Computer Vision`
+
+Gyro Dropout replaces implementations of `torch.nn.Dropout`. The Gyro Dropout provides increased accuracy compared with dropout.
+
+Gyro dropout is a variant of dropout that improves the efficiency of training neural networks
+Instead of randomly dropping out neurons in every training iteration, gyro dropout pre-selects and trains a fixed
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:11]:
```suggestion
Gyro dropout is a variant of dropout that improves the efficiency of training neural networks.
Instead of randomly dropping out neurons in every training iteration, gyro dropout pre-selects and trains a fixed
```

File: composer/algorithms/gyro_dropout/README.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+number of subnetwork. 'Sigma' is the number of total pre-selected subnetworks and 'Tau' is the number of concurrently scheduled subnetworks in an iteration
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:12]:
```suggestion
number of subnetworks. 'Sigma' is the number of total pre-selected subnetworks and 'Tau' is the number of concurrently scheduled subnetworks in an iteration.
```

File: composer/algorithms/gyro_dropout/README.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                  algorithms=[GyroDropout(p=0.5,
+                                          sigma=256,
+                                          tau=16)])
+
+trainer.fit()
+```
+
+### Implementation Details
+
+Gyro Dropout is implemented by performing model surgery, which looks for instances of `torch.nn.Dropout`. This should be applicable to any model that utilizes a `torch.nn.Dropout`.
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:64]:
```suggestion
Gyro Dropout is implemented by performing model surgery, which looks for instances of `torch.nn.Dropout`. This should be applicable to any model that utilizes `torch.nn.Dropout`.
```

File: composer/algorithms/gyro_dropout/README.md (hunk 5 above)
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:70]:
```suggestion
For the hyperparameters (sigma, tau), we recommend (256, 16) in AlexNet, LeNet or (1024, 8) in ResNet-18, BERT.
```

File: composer/algorithms/gyro_dropout/README.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+## Suggested Hyperparameters
+
+Gyro Dropout has two hyperparameters - sigma, tau. (iters_per_epoch and max_epoch is training-dependent)
+
+We recommend (256, 16) in AlexNet, LeNet or (1024, 8) in ResNet-18, BERT as hyperparameter set for (sigma, tau).
+
+But there might be better possible (sigma, tau) set that may achieve higher accuracy for other models.
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:72]:
```suggestion
```

Comment (mvpatel2000) [reply composer/algorithms/gyro_dropout/README.md:72]: I think it's probably fine to leave this off and only give recommendations for what you have data on.

File: composer/algorithms/gyro_dropout/README.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+## Technical Details
+GyroDropout achieves improved accuracy over conventional dropout by pre-selecting a fixed number of subnetworks and training with only those subnetworks. Because the selected subnetworks are trained more robustly (compared to the conventional dropout cases), their diversity increases and thus their ensemble achieves higher accuracy.
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:75]:
```suggestion
GyroDropout achieves improved accuracy over conventional dropout by pre-selecting a fixed number of subnetworks and training with only those subnetworks. Because the selected subnetworks are trained more robustly (compared to the conventional dropout), their diversity increases and thus their ensemble achieves higher accuracy.
```

File: composer/algorithms/gyro_dropout/gyro_dropout.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+from composer.algorithms.warnings import NoEffectWarning
+from composer.core import Algorithm, Event, State
+from composer.loggers import Logger
+from composer.utils import module_surgery
+
+log = logging.getLogger(__name__)
+
+
+class GyroDropoutLayer(torch.nn.Module):
+    def __init__(self, iters_per_epoch: int, max_epoch: int, p:float, sigma:int, tau:int):
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/gyro_dropout.py:19]:
```suggestion
    def __init__(self, iters_per_epoch: int, max_epoch: int, p: float, sigma: int, tau: int):
```

File: composer/functional/__init__.py [hunk 12]
```diff
@@ -67,4 +68,5 @@
     'apply_squeeze_excite',
     'apply_stochastic_depth',
     'apply_weight_standardization',
+    'apply_gyro_dropout,
```
Comment (mvpatel2000) [inline composer/functional/__init__.py:71]:
```suggestion
    'apply_gyro_dropout',
```

**Review by mvpatel2000 — COMMENTED** [review]
Apologies for the delay! Daya is on paternity leave, and we were figuring out the right person to take this on internally. This mostly looks good to me -- I've left a couple minor wording suggestions.

The only remaining ask -- it would be great if you could please include some unit tests to ensure functionality. For examples, take a look at what we've done for other algorithms. After that, we'd love to merge this in and advertise your work!

**Comment by jelite** [comment]
> Apologies for the delay! Daya is on paternity leave, and we were figuring out the right person to take this on internally. This mostly looks good to me -- I've left a couple minor wording suggestions.
> 
> The only remaining ask -- it would be great if you could please include some unit tests to ensure functionality. For examples, take a look at what we've done for other algorithms. After that, we'd love to merge this in and advertise your work!

**Comment by jelite** [comment]
> Apologies for the delay! Daya is on paternity leave, and we were figuring out the right person to take this on internally. This mostly looks good to me -- I've left a couple minor wording suggestions.
> 
> The only remaining ask -- it would be great if you could please include some unit tests to ensure functionality. For examples, take a look at what we've done for other algorithms. After that, we'd love to merge this in and advertise your work!


Hello! I have added some unit tests in test_gyro_dropout.py. If you have any problems, please feel free to add some comments.

**Review by mvpatel2000 — COMMENTED** [review]
@jelite this looks mostly good to me! Can you please make sure the tests pass and then I can give you approval? 

It looks like you have some lint issues. https://github.com/mosaicml/composer/blob/dev/STYLE_GUIDE.md gives some instructions for how to fix them

File: tests/algorithms/algorithm_settings.py [hunk 13]
```diff
@@ -150,6 +150,7 @@
         }
     },
     WeightStandardization: simple_vision_settings,
+    GyroDropout: None, #Dropout settings needed
```
Comment (mvpatel2000) [inline tests/algorithms/algorithm_settings.py:153]: This will put xfail around any tests that use GyroDropout form algorithm settings. 
```suggestion
    GyroDropout: None,  # Dropout settings needed
```

Comment (mvpatel2000) [reply tests/algorithms/algorithm_settings.py:153]: Minor style nit to be consistent

**Comment by jelite** [comment]
@mvpatel2000 I have found some pyright errors and edited code to pass pyright tests. After that, I checked my code passes pyright tests that have some errors before. But it couldn't pass Jenkins, and I can't check what problem is.  So, could you push directly to this branch and fix them? Thank you for considering my work.

File: composer/algorithms/gyro_dropout/README.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
+# 🍰 Gyro Dropout
+
+
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:3]:
```suggestion

```

File: docs/source/method_cards/gyro_dropout.md [hunk 15]
```diff
@@ -0,0 +1 @@
+../../../composer/algorithms/gyro_dropout/README.md
```
Comment (mvpatel2000) [inline docs/source/method_cards/gyro_dropout.md:1]: This file should actually be a symbolic link to the README. Inside the folder `/docs/source/method_cards/`, @jelite  can you please run `ln -s ../../../composer/algorithms/gyro_dropout/README.md gyro_dropout.md`

**Review by mvpatel2000 — COMMENTED** [review]
@jelite  it appears I do not have the right permission set to be able to push. I've left a comment describing the reason why the test is failing -- the doc should be a symbolic link. Would you mind doing the change please? and then we should be good to merge! 🎉

File: composer/algorithms/gyro_dropout/README.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+trainer = Trainer(model=model,
+                  train_dataloader=train_dataloader,
+                  eval_dataloader=eval_dataloader,
+                  max_duration='100ep',
+                  algorithms=[GyroDropout(p=0.5,
+                                          sigma=256,
+                                          tau=16)])
+
+trainer.fit()
+```
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:60]: Before this, the following snippet needs to be inserted:


\


to setup the doctest. This is done in other algorithm READMEs that you can take a look at. This is so these code snippets can successfully be autotested (its one of the current test failures)

File: composer/algorithms/gyro_dropout/README.md [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+# Apply surgery on the model to swap-in the Gyro Dropout using the Composer functional API
+
+import composer.functional as cf
+
+def training_loop(model, train_loader):
+    cf.apply_gyro_dropout(
+        model,
+        iters_per_epoch = 196,
+        max_epoch = 100,
+        p = 0.5
```
Comment (mvpatel2000) [inline composer/algorithms/gyro_dropout/README.md:28]:
```suggestion
        p = 0.5,
```

**Review by mvpatel2000 — COMMENTED** [review]
@jelite  it looks like the tests for the READMEs are failing. I've put fixes for the tests in the PR review. Can you please make the tweaks so they pass? Should be the last changes :)

File: composer/algorithms/gyro_dropout/README.md (hunk 16 above)
Comment (mvpatel2000) [reply composer/algorithms/gyro_dropout/README.md:60]: example from `label_smoothing`: 
image

**Review by mvpatel2000 — APPROVED** [review]
Looks good to me! Thank you all for your hard work on this. Please feel free to merge whenever

**Comment by jelite** [comment]
> Looks good to me! Thank you all for your hard work on this. Please feel free to merge whenever

@mvpatel2000 I merged branch 'dev' into gyro_dropout. Thank you for your help of this work!
